File: s_coding/삼성 SW 역량 테스트 기출 문제/14501_퇴사.cpp
#if 01
//퇴사
//https://www.acmicpc.net/problem/14501
#include <iostream>
using namespace std;

int N,sol;
int t[15 + 1]; //일하는데 걸리는 기간
int p[15 + 1]; //일하면 받는 돈

void Solve(int day,int total) {

	if (day == N + 1) { //ex) 8일이되면 , 계산조건 
		if (sol < total)sol = total;
		return;
	}
	if (day > N + 1)return; //ex) 8일보다 커지면 종료 , 종료조건
	//if(day + t[day] <= N+ 1)
	//if(day + 1 <= N + 1)

	//상담하는 경우
	if (day + t[day] <= N + 1)Solve(day + t[day], total + p[day]);

	//상담하지 않는 경우
	if (day + 1 <= N + 1)Solve(day + 1, total);
}
void InputData() {
	scanf("%d", &N);

	for (int i = 1; i <= N; i++) {
		scanf("%d %d", &t[i], &p[i]);
	}
}
int main() {

	freopen("input.txt", "r", stdin);
	InputData();
	Solve(1,0);

	printf("%d", sol);
	return 0;
}
#endif

File: s_coding/삼성 SW 역량 테스트 기출 문제/14501-퇴사.py
#퇴사
#https://www.acmicpc.net/problem/14501
import sys
sys.stdin = open("input.txt","r")

n=int(input())
t=[] #걸리는 시간
p=[] #수익
res=-1e9
def dfs(day,sum):
    global res
    if day==n+1:
        if sum > res:
            res=sum
    else:
        if day+t[day]<=n+1: #상담이 가능한 경우
            dfs(day+t[day],sum+p[day])
        dfs(day+1,sum)

for i in range(n):
    a,b=map(int,input().split())
    t.append(a) 
    p.append(b)
t.insert(0,0) #인덱스번호를 날짜로 사용하기 위해
p.insert(0,0)
dfs(1,0)
print(res)
